File: simplecompiler/compiler/parser.py
#! /usr/bin/env python3
"""Parser for simplecompiler"""
# Original copyright:
# Copyright 2004-2005 Elemental Security, Inc. All Rights Reserved.
# Licensed to PSF under a Contributor Agreement.

import logging
import pickle

from pathlib import Path
from pprint import pprint
from collections import namedtuple

from simplecompiler.compiler.Symbol import ERRORTOKEN
from simplecompiler.compiler.tokenizer import is_terminal, is_nonterminal


__all__ = ["load_grammar", "Parser", "lispify", "pprint_cst"]

logger = logging.getLogger()
logging.basicConfig(level=logging.INFO)


def load_grammar():
    """Load Grammar.pickle from simplecompiler/compiler"""
    with (Path(__file__).parent / "Grammar.pickle").open('rb') as f:
        return pickle.load(f)


# A node is a tuple: (type, value, context, children).
# where children is a list of nodes or None, and context may be None.
# context is actually the source location as a 2-tuple.
class Node(namedtuple('Node', 'type value context children')):
    "A Node in the concrete syntax tree."""

    @property
    def first_child(self):
        return self.children[0]


# Each stack entry is a tuple: (dfa, state, node).
StackEntry = namedtuple('StackEntry', 'dfa state node')


class Parser:
    """Parser engine."""

    def raise_error(self, msg, token):
        args = (self.filename, *token.start, token.line)
        raise SyntaxError(msg, args)

    def __init__(self, grammar, filename):

        self.grammar = grammar
        # for report error
        self.filename = filename
        start = grammar.start
        newnode = Node(start, None, None, [])
        stackentry = StackEntry(self.grammar.dfas[start], 0, newnode)
        self.stack = [stackentry]
        self.rootnode = None

    def addtoken(self, token):
        """Add a token; return True iff this is the end of the program."""
        logger.debug('addtoken: {}'.format(token))
        type, value, context, *_ = token
        # Map from token to label
        ilabel = self.classify(type, value, context)
        # Loop until the token is shifted; may raise exceptions
        assert self.stack, (type, value)
        while True:
            dfa, state, node = self.stack[-1]
            states, first = dfa
            arcs = states[state]
            # Look for a state with this label
            for i, newstate in arcs:
                t, v = self.grammar.labels[i]
                if ilabel == i:
                    # print("shift", self.grammar.tok_name[t], value)
                    # Look it up in the list of labels
                    assert is_terminal(t)
                    # Shift a token; we're done with it
                    self.shift(type, value, newstate, context)
                    # Pop while we are in an accept-only state
                    state = newstate
                    while states[state] == [(0, state)]:
                        self.pop()
                        if not self.stack:
                            # Done parsing!
                            return True
                        dfa, state, node = self.stack[-1]
                        states, first = dfa
                    # Done with this token
                    return False
                elif is_nonterminal(t):
                    # See if it's a symbol and if we're in its first set
                    itsdfa = self.grammar.dfas[t]
                    itsstates, itsfirst = itsdfa
                    if ilabel in itsfirst:
                        # print("push", self.grammar.number2symbol[t])
                        # Push a symbol
                        self.push(t, self.grammar.dfas[t], newstate, context)
                        break  # To continue the outer while loop
            else:
                if (0, state) in arcs:
                    # An accepting state, pop it and try something else
                    self.pop()
                    if not self.stack:
                        # Done parsing, but another token is input
                        self.raise_error("too much input", token)
                else:
                    # No success finding a transition
                    self.raise_error("unexpected {!r}".format(value), token)

    def classify(self, type, value, context):
        """Turn a token into a label.  (Internal)"""
        if value in self.grammar.keywords:
            # Check for reserved words
            ilabel = self.grammar.keywords.get(value)
            if ilabel is not None:
                return ilabel
        ilabel = self.grammar.tokens.get(type)
        if ilabel is None:
            raise SyntaxError("unexpected {!r}".format(value))
            # raise ParseError("bad token", type, value, context)
        return ilabel

    def shift(self, type, value, newstate, context):
        """Shift a token.  (Internal)"""
        logger.debug('shift')
        dfa, state, node = self.stack[-1]
        newnode = Node(type, value, context, None)
        node.children.append(newnode)
        self.stack[-1] = StackEntry(dfa, newstate, node)

    def push(self, type, newdfa, newstate, context):
        """Push a nonterminal.  (Internal)"""
        logger.debug('push')
        dfa, state, node = self.stack[-1]
        newnode = Node(type, None, context, [])
        self.stack[-1] = StackEntry(dfa, newstate, node)
        self.stack.append(StackEntry(newdfa, 0, newnode))

    def pop(self):
        """Pop a nonterminal.  (Internal)"""
        # print("pop")
        *_, newnode = self.stack.pop()
        if self.stack:
            dfa, state, node = self.stack[-1]
            node.children.append(newnode)
        else:
            self.rootnode = newnode

    # public interface
    def parse_tokens(self, tokens):
        """Parse a series of tokens and return the concrete syntax tree.
        On error, SyntaxError is raised.
        """
        for token in tokens:
            if token.type == ERRORTOKEN:
                self.raise_error("bad token", token)
            if self.addtoken(token):
                break
        else:
            self.raise_error("incomplete input", token)
        return self.rootnode


def lispify(cst_node, grammar):
    """Turn cst_node into a nested tuple so that it can be
    pretty printed with proper indents.

    Note although cst_node is implemented as namedtuple, there are
    None fields within each node and they store symbol numbers rather
    than names. This function removes None fields and turns numbers into names
    so that the resulting cst_node becomes more readable.
    """
    out = []
    type, *rest, children = cst_node
    if is_terminal(type):
        type = grammar.tok_name[type]
    else:
        # nonterminal
        type = grammar.number2symbol[type]
    out.append(type)
    out.extend(filter(None, rest))
    if children is not None:
        children = tuple(lispify(child, grammar) for child in children)
        out.append(children)
    return tuple(out)


def pprint_cst(cst_node, grammar, file):
    """Pretty print an cst_node to file"""
    return pprint(lispify(cst_node, grammar), file)

File: Codegen/parser.cpp
#include "parser.h"
#include "error.h"
#include "Node.h"

#include <cstdio>
#include <sstream>
#include <stack>
#include <vector>

using namespace simplecompiler;

Node::~Node() {
  for (auto child : children)
    delete child;
}

String Node::FormatValue() const {
  return type != Symbol::ENDMARKER && value.empty() ? "None" : Quote(value);
}

void Node::Format(std::ostream &os) const {
  const char *type_str = GetSymName(type);
  os << "Node(";
  os << "type=" << type_str << ", ";
  os << "value=" << FormatValue() << ", ";
  os << "context=" << location << ", ";
  os << "children=";
  if (children.empty()) {
    os << "None)";
  } else {
    os << "[";
    for (int i = 0; i < children.size(); i++) {
      children[i]->Format(os);
      if (i != children.size() - 1)
        os << ", ";
    }
    os << "])";
  }
}

namespace {
class StackEntry {
public:
  DFA *dfa;
  int state;
  Node *node;

  StackEntry(DFA *dfa, int state, Node *node)
      : dfa(dfa), state(state), node(node) {}
};

bool IsInFirst(DFA *dfa, int label) {
  for (int i = 0; i < dfa->n_first; i++)
    if (label == dfa->first[i])
      return true;
  return false;
}

bool IsAcceptOnlyState(DFAState *state) {
  return state->is_final && state->n_arcs == 1;
}

void DumpStackEntry(StackEntry e) {
  printf("state: %d\n", e.state);
  printf("dfa: %s\n", e.dfa->name);
  /* DumpDFA(e.dfa); */
}

class Parser {
public:
  std::stack<StackEntry> stack;
  Grammar *grammar;
  Node *rootnode;
  ErrorManager e;

  explicit Parser(Grammar *grammar)
      : stack(), grammar(grammar), rootnode(nullptr), e() {
    auto start = grammar->start;
    Node *newnode = new Node(static_cast<Symbol>(start), "", Location(0, 0));
    stack.push(StackEntry(grammar->dfas[start - NT_OFFSET], 0, newnode));
  }

  int Classify(const TokenInfo &token) {
    if (token.type == Symbol::NAME || token.type == Symbol::OP) {
      for (int i = 1; i < grammar->n_labels; i++) {
        const Label &l = grammar->labels[i];
        if (l.string && l.string == token.string)
          return i;
      }
    }
    for (int i = 1; i < grammar->n_labels; i++) {
      const Label &l = grammar->labels[i];
      if (l.type == static_cast<int>(token.type) && l.string == nullptr) {
        return i;
      }
    }
    e.SyntaxError(token.start, "unexpected", Quote(token.string));
    return -1;
  }

  void Shift(const TokenInfo &token, int newstate) {
    /* printf("shift %s\n", GetSymName(token.type)); */
    StackEntry &tos = stack.top();
    tos.node->AddChild(new Node(token.type, token.string, token.start));
    tos.state = newstate;
  }

  void Push(Symbol type, DFA *newdfa, int newstate, Location location) {
    /* printf("push %s\n", GetSymName(type)); */
    StackEntry &tos = stack.top();
    Node *newnode = new Node(type, "", location);
    tos.state = newstate;
    stack.push(StackEntry(newdfa, 0, newnode));
  }

  void Pop() {
    /* printf("pop\n"); */
    StackEntry tos = stack.top();
    stack.pop();
    Node *newnode = tos.node;

    if (stack.size()) {
      stack.top().node->AddChild(newnode);
    } else {
      rootnode = newnode;
    }
  }

  bool AddToken(const TokenInfo &token) {
    int label = Classify(token);
    if (label < 0) {
      return -1;
    }
    /* token.Format(stdout); */

    while (true) {
      StackEntry &tos = stack.top();
      DFA *dfa = tos.dfa;
      DFAState *states = dfa->states;
      DFAState *state = &states[tos.state];
      bool flag = true;

      /* DumpStackEntry(tos); */

      for (int i = 0; i < state->n_arcs; ++i) {
        Arc &arc = state->arcs[i];
        auto type = static_cast<Symbol>(grammar->labels[arc.label].type);
        int newstate = arc.state;

        if (label == arc.label) {
          Shift(token, newstate);
          /* printf("shift %s\n", GetSymName(type)); */

          while (IsAcceptOnlyState(&states[newstate])) {
            /* printf("pop\n"); */
            Pop();
            if (stack.empty()) {
              return true;
            }
            newstate = stack.top().state;
            states = stack.top().dfa->states;
          }
          return false;
        }

        else if (IsNonterminal(type)) {
          DFA *itsdfa = grammar->dfas[static_cast<int>(type) - NT_OFFSET];
          if (IsInFirst(itsdfa, label)) {
            /* printf("push %s\n", GetSymName(type)); */
            Push(type, itsdfa, newstate, token.start);
            flag = false;
            break;
          }
        }
      }

      if (flag) {
        if (state->is_final) {
          Pop();
          if (stack.empty()) {
            e.SyntaxError(token.start, "too much input");
            return -1;
          }
        } else {
          e.SyntaxError(token.start, "unexpected", Quote(token.line));
          return -1;
        }
      }
    }
  }

  Node *ParseTokens(const TokenBuffer &tokens) {
    for (const auto &token : tokens) {
      if (token.type == Symbol::ERRORTOKEN) {
        e.SyntaxError(token.start, "error token", Quote(token.string));
        return nullptr;
      }
      int ret = AddToken(token);
      if (ret == 1) {
        return rootnode;
      }
      if (ret < 0) {
        return nullptr;
      }
    }
    auto last = tokens.end() - 1;
    e.SyntaxError(last->start, "incomplete input");
    return nullptr;
  }
};
} // namespace

Node *simplecompiler::ParseTokens(const TokenBuffer &tokens) {
  Parser parser(&CompilerGrammar);
  return parser.ParseTokens(tokens);
}
